catroid/src/main/java/org/catrobat/catroid/stage/TextToSpeechHolder.kt

```diff
@@ -0,0 +1,103 @@
+ * http://developer.catrobat.org/license_additional_term
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.catrobat.catroid.stage
+
+import android.content.Intent
+import android.speech.tts.TextToSpeech
+import android.speech.tts.TextToSpeech.OnInitListener
+import android.speech.tts.UtteranceProgressListener
+import android.util.Log
+import android.view.ContextThemeWrapper
+import androidx.appcompat.app.AlertDialog
+import org.catrobat.catroid.R
+import org.catrobat.catroid.common.Constants
+import org.catrobat.catroid.content.bricks.Brick
+import java.io.File
+import java.util.HashMap
+
+class TextToSpeechHolder private constructor() {
+
+    fun initTextToSpeech(stageActivity: StageActivity, stageResourceHolder: StageResourceHolder) {
+        textToSpeech = TextToSpeech(stageActivity, OnInitListener { status ->
+                if (status == TextToSpeech.SUCCESS) {
+                    utteranceProgressListenerContainer = UtteranceProgressListenerContainer()
+                    textToSpeech?.setOnUtteranceProgressListener(utteranceProgressListenerContainer)
+                    stageResourceHolder.resourceInitialized()
+                } else {
+                    val builder = AlertDialog.Builder(ContextThemeWrapper(stageActivity, R.style.Theme_AppCompat_Dialog))
+                    builder.setMessage(R.string.prestage_text_to_speech_engine_not_installed)
+                        .setCancelable(false)
+                        .setPositiveButton(R.string.yes) { dialog, id ->
+                            val installIntent = Intent()
+                            installIntent.action = TextToSpeech.Engine.ACTION_INSTALL_TTS_DATA
+                            stageActivity.startActivity(installIntent)
+                            stageResourceHolder.resourceFailed(Brick.TEXT_TO_SPEECH)
+                        }
+                        .setNegativeButton(R.string.no) { dialog, id ->
+                            stageResourceHolder.resourceFailed(Brick.TEXT_TO_SPEECH)
+                        }
+                    val alert = builder.create()
+                    alert.show()
+                }
+            })
+    }
+
+    fun shutDownTextToSpeech() {
+        if (textToSpeech != null) {
+            textToSpeech?.stop()
+            textToSpeech?.shutdown()
+        }
+    }
+
+    fun textToSpeech(text: String?, speechFile: File, listener: UtteranceProgressListener, speakParameter: HashMap<String?, String?>) {
+        val text = text ?: ""
+        if (utteranceProgressListenerContainer!!.addUtteranceProgressListener(speechFile,
```

Comment:
Never use "!!" it's an "assert not null". if the value is null => Process crash. 
Do we really want a NPE here?
Have a look at https://kotlinlang.org/docs/reference/null-safety.html#the--operator